pilot/pkg/networking/core/v1alpha3/listener_builder.go

```diff
@@ -501,44 +501,67 @@ func buildInboundCatchAllHTTPFilterChains(configgen *ConfigGeneratorImpl,
 			Bind:                       matchingIP,
 			InboundClusterName:         clusterName,
 		}
+		// Call plugins to install authn/authz policies.
+		var allChains []istionetworking.FilterChain
+		for _, p := range configgen.Plugins {
+			chains := p.OnInboundPassthroughFilterChains(in)
+			allChains = append(allChains, chains...)
+		}
+		if len(allChains) == 0 {
+			// Add one empty entry to the list if none of the plugins are interested in updating the filter chains.
+			allChains = []istionetworking.FilterChain{{}}
+		}
+
+		// Override the filter chain match to make sure the pass through filter chain captures the pass through traffic.
+		for i := range allChains {
+			chain := &allChains[i]
+			if chain.FilterChainMatch == nil {
+				chain.FilterChainMatch = &listener.FilterChainMatch{}
+			}
+			chain.FilterChainMatch.PrefixRanges = []*core.CidrRange{
+				util.ConvertAddressToCidr(matchingIP),
+			}
+			chain.FilterChainMatch.ApplicationProtocols = plaintextHTTPALPNs
+			chain.ListenerProtocol = istionetworking.ListenerProtocolHTTP
+		}
+
 		mutable := &istionetworking.MutableObjects{
-			FilterChains: []istionetworking.FilterChain{
-				{
-					ListenerProtocol: istionetworking.ListenerProtocolHTTP,
-				},
-			},
+			FilterChains: allChains,
 		}
 		for _, p := range configgen.Plugins {
 			if err := p.OnInboundPassthrough(in, mutable); err != nil {
 				log.Errorf("Build inbound passthrough filter chains error: %v", err)
 			}
 		}
-		httpOpts := configgen.buildSidecarInboundHTTPListenerOptsForPortOrUDS(node, in)
-		httpOpts.statPrefix = clusterName
-		connectionManager := buildHTTPConnectionManager(in, httpOpts, mutable.FilterChains[0].HTTP)
 
-		filter := &listener.Filter{
-			Name:       xdsutil.HTTPConnectionManager,
-			ConfigType: &listener.Filter_TypedConfig{TypedConfig: util.MessageToAny(connectionManager)},
-		}
+		// Construct the actual filter chains for each of the filter chain from the plugin.
+		for _, chain := range allChains {
+			httpOpts := configgen.buildSidecarInboundHTTPListenerOptsForPortOrUDS(node, in)
+			httpOpts.statPrefix = clusterName
+			connectionManager := buildHTTPConnectionManager(in, httpOpts, chain.HTTP)
 
-		filterChainMatch := listener.FilterChainMatch{
-			// Port : EMPTY to match all ports
-			PrefixRanges: []*core.CidrRange{
-				util.ConvertAddressToCidr(matchingIP),
-			},
-			ApplicationProtocols: plaintextHTTPALPNs,
-		}
+			filter := &listener.Filter{
+				Name:       xdsutil.HTTPConnectionManager,
+				ConfigType: &listener.Filter_TypedConfig{TypedConfig: util.MessageToAny(connectionManager)},
+			}
 
-		filterChain := &listener.FilterChain{
-			Name:             VirtualInboundListenerName,
-			FilterChainMatch: &filterChainMatch,
-			Filters: []*listener.Filter{
-				filter,
-			},
+			filterChain := &listener.FilterChain{
+				FilterChainMatch: chain.FilterChainMatch,
+				Filters:          append(chain.TCP, filter),
```

Comment:
TCP? shouldn't this be HTTP?

Reply:
This is the network filter. HTTP filter was applied before.

Reply:
I'm a little confused here, could you give a high-level structure of the generated filter chain?

Reply:
The filter chain includes multiple network layer (L4) filters and HTTP connection manager (HCM) is one of the L4 filters. The HCM includes multiple L7 filters specific for HTTP protocols.

Reply:
I am not understanding this either. Are you trying to say we should generate a single filter chain with `tcpProxy filter` AND `HCM`? shouldn't this be TCP or HCM?

Reply:
This should be TCP. HCM is appended
```go
// TCP is the set of network (TCP) filters for this filter chain.
TCP []*listener.Filter
```
The filters include both HCM and other network layer(L4) filters.

Reply:
Can you give an example of a l4 filter that we want to happen before HCM?

Reply:
I didn't have example on hand. I could find some comment from existing code base. https://github.com/istio/istio/blame/master/pilot/pkg/networking/core/v1alpha3/listener.go#L2363

Maybe it's out of date?

Reply:
The only examples I can find is adding a tcpproxy filter, which will break this, since the tcpproxy filter is terminating. Which leads me to believe this may not be quite right?
If we don't understand what we are doing that is a sign its not tested enough

Reply:
The code is trying to add the l4 filter before the HCM. TCPproxy won't be appended here since only the l4 proxy generated by the plugin will be appended here. I could find some l4 filters like metadata exchanged was added before the HCM.

Reply:
Can we please see an explicit example of a case where `chain.TCP` is non-empty and this is the correct behavior? I still can't wrap my head around how this can be correct and I don't want us guessing.